### PR TITLE
[main][feat] Automated Vaults Repurchase

### PR DESCRIPTION
**Pull request by alpanaca**
## Description
New version of DeltaNeutralVault that allow repurchase process.

## Why?
To reduce price impact during rebalancing.

## ChangeLogs:
- two new strats specific for repurchase
- delta04.repurchase()
- test files

### CODE REVIEW

File: solidity/contracts/6/protocol/strategies/repurchase/RepurchaseBorrowStrategy.sol [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import "@openzeppelin/contracts-ethereum-package/contracts/utils/ReentrancyGuard.sol";
+import "@openzeppelin/contracts-ethereum-package/contracts/Initializable.sol";
+import "@openzeppelin/contracts-ethereum-package/contracts/access/Ownable.sol";
+
+import "../../interfaces/IPancakePair.sol";
+import "../../interfaces/IStrategy.sol";
+import "../../interfaces/IVault.sol";
+import "../../interfaces/IWorker.sol";
+
+import "../../apis/pancake/IPancakeRouter02.sol";
```
Comment (alpanaca) [inline solidity/contracts/6/protocol/strategies/repurchase/RepurchaseBorrowStrategy.sol:26]: Should remove this

File: solidity/contracts/6/protocol/strategies/repurchase/RepurchaseBorrowStrategy.sol [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  mapping(address => bool) public okWorkers;
+  mapping(address => bool) public okDeltaNeutralVaults;
+
+  /// @notice require that only allowed workers are able to do the rest of the method call
+  modifier onlyWhitelistedWorkers() {
+    require(okWorkers[msg.sender], "RepurchaseBorrowStrategy::onlyWhitelistedWorkers:: bad worker");
+    _;
+  }
+
+  /// @dev Create a new add two-side optimal strategy instance.
```
Comment (jr-alpaca) [inline solidity/contracts/6/protocol/strategies/repurchase/RepurchaseBorrowStrategy.sol:40]:
```suggestion
  /// @dev Create a new repurchase borrow strategy instance.
```

File: solidity/contracts/8.13/DeltaNeutralVault04.sol [hunk 3]
```diff
@@ -500,6 +510,51 @@ contract DeltaNeutralVault04 is IDeltaNeutralStruct, ERC20Upgradeable, Reentranc
...
+      if (_tokenIn != assetToken) revert DeltaNeutralVault04_InvalidRepurchaseTokenIn();
+    }
+
+    address _tokenOut = _tokenIn == stableToken ? assetToken : stableToken;
+
+    _amountOut = FullMath.mulDiv(
+      _amountIn,
+      FullMath.mulDiv(_getTokenPrice(_tokenOut), 1e18, _getTokenPrice(_tokenIn)),
+      1e18
+    );
```
Comment (jr-alpaca) [inline solidity/contracts/8.13/DeltaNeutralVault04.sol:531]: Token decimals of stable and asset token may not be the same. Should normalize both token amount before calculation.

File: solidity/contracts/8.13/DeltaNeutralVault04.sol [hunk 4]
```diff
@@ -500,6 +510,51 @@ contract DeltaNeutralVault04 is IDeltaNeutralStruct, ERC20Upgradeable, Reentranc
...
+
+    if (_amountOut > uint256(_assetExposure >= 0 ? _assetExposure : (_assetExposure * -1)))
```
Comment (jr-alpaca) [inline solidity/contracts/8.13/DeltaNeutralVault04.sol:533]: _amountOut may be in the form of baseToken which we cannot use to compare with assetExposure.

**Comment by alpanaca** [comment]
Repurchase should limit to EOA only.

File: solidity/contracts/6/protocol/strategies/repurchase/RepurchaseBorrowStrategy.sol [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  /// @param data Extra calldata information passed along to this strategy.
+  function execute(
+    address, /* user */
+    uint256,
+    /* debt */
+    bytes calldata data
+  ) external override onlyWhitelistedWorkers nonReentrant {
+    // 1. Retrieve the DeltaNeutralVault which perform this repurchasing
+    address _deltaNeutralVault = abi.decode(data, (address));
+    require(okDeltaNeutralVaults[_deltaNeutralVault], "RepurchaseBorrowStrategy::execute:: bad target");
```
[the remark this answers is not in the document]
Comment (alpanaca) [reply solidity/contracts/6/protocol/strategies/repurchase/RepurchaseBorrowStrategy.sol:56]: I think this is a safety net for Executor that compose the input which is out of our scope. Can leave it this way.

**Review by jr-alpaca — DISMISSED** [review]

File: deploy/exec/strategies/config/set-whitelisted-workers.ts [hunk 6]
```diff
@@ -1,16 +1,15 @@
-import { HardhatRuntimeEnvironment } from 'hardhat/types';
-import { DeployFunction } from 'hardhat-deploy/types';
-import { ethers } from 'hardhat';
-import { Timelock__factory } from '../../../../typechain'
-
+import { HardhatRuntimeEnvironment } from "hardhat/types";
```
Comment (abc-alpaca) [inline deploy/exec/strategies/config/set-whitelisted-workers.ts:1]: merge logic with `set-whitelisted-workers-timelock`

File: package.json [hunk 7]
```diff
@@ -498,7 +498,7 @@
     "test:av-controller": "forge test --match-path solidity/tests/protocol/private-automated-vault/AutomatedVaultController.t.sol",
     "test:delta02": "forge test --match-path solidity/tests/protocol/private-automated-vault/DeltaNeutralVault02.t.sol",
     "test:directional-vault": "forge test --match-path solidity/tests/protocol/directional-vault/DirectionalVault.t.sol",
-    "test:delta04": "forge test --match-path solidity/tests/protocol/delta-neutral-04/DeltaNeutralVault04.t.sol"
+    "test:delta04": "forge test --match-contract DeltaNeutralVault04"
```
Comment (abc-alpaca) [inline package.json:501]: need command like this
```
"deploy:mainnet:strategies:config:set-whitelisted-delta" : "hardhat --network mainnet deploy --no-compile --reset --tags SetSharedStratsWhitelistedDeltaNeutralVaults",
"deploy:mainnet:strategies:config:set-whitelisted-worker" : "hardhat --network mainnet deploy --no-compile --reset --tags SetSharedStratsWhitelistedWorkers",
"deploy:mainnet:strategies:deploy:repurchase-borrow":  : "hardhat --network mainnet deploy --no-compile --reset --tags RepurchaseBorrowStrategies",
"deploy:mainnet:strategies:deploy:repurchase-repay":  : "hardhat --network mainnet deploy --no-compile --reset --tags RepurchaseRepayStrategies",

```

**Review by abc-alpaca — CHANGES_REQUESTED** [review]

**Review by sora-alpaca — DISMISSED** [review]

**Review by abc-alpaca — CHANGES_REQUESTED** [review]

**Review by jr-alpaca — APPROVED** [review]

**Review by ricardo-alpaca — DISMISSED** [review]

**Review by almost-llamas — APPROVED** [review]

**Review by jr-alpaca — APPROVED** [review]

**Pull request merged**